tests/functional/test_s3.py

```diff
@@ -38,13 +40,81 @@ def test_transfer_methods_injected_to_bucket(self):
+        self.assertTrue(hasattr(bucket, 'copy'),
+                        'copy was not injected onto S3 bucket')
 
     def test_transfer_methods_injected_to_object(self):
         obj = boto3.resource('s3').Object('my_bucket', 'my_key')
         self.assertTrue(hasattr(obj, 'upload_file'),
                         'upload_file was not injected onto S3 object')
         self.assertTrue(hasattr(obj, 'download_file'),
                         'download_file was not injected onto S3 object')
+        self.assertTrue(hasattr(obj, 'copy'),
+                        'copy was not injected onto S3 object')
+
+
+class TestCopy(unittest.TestCase):
+    def setUp(self):
+        self.session = boto3.session.Session(
+            aws_access_key_id='foo', aws_secret_access_key='bar',
+            region_name='us-west-2')
+        self.s3 = self.session.resource('s3')
+        self.stubber = Stubber(self.s3.meta.client)
+        self.add_success()
+        self.copy_source = {'Bucket': 'foo', 'Key': 'bar'}
+
+    def add_success(self):
+        head_response = {
+            'AcceptRanges': 'bytes',
+            'ContentLength': 4,
+            'ContentType': 'binary/octet-stream',
+            'ETag': '"example0etag"',
+            'Metadata': {},
+            'ResponseMetadata': {
+                'HTTPHeaders': {
+                    'accept-ranges': 'bytes',
+                    'content-length': '4',
+                    'content-type': 'binary/octet-stream',
+                    'date': 'Fri, 15 Jul 2016 20:14:03 GMT',
+                    'etag': '"example0etag"',
+                    'last-modified': 'Tue, 12 Jul 2016 21:26:07 GMT',
+                    'server': 'AmazonS3',
+                },
+                'HTTPStatusCode': 200,
+            }
+        }
+        copy_response = {
+            'CopyObjectResult': {
+                'ETag': '"example0etag"'
+            },
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200
+            }
+        }
+        self.stubber.add_response(
+            method='head_object', service_response=head_response)
+        self.stubber.add_response(
+            method='copy_object', service_response=copy_response)
+
+    def test_client_copy(self):
+        with self.stubber:
+            response = self.s3.meta.client.copy(
+                self.copy_source, 'mybucket', 'mykey')
+        # The response will be none on a successful transfer.
+        self.assertIsNone(response)
+
+    def test_bucket_copy(self):
+        bucket = self.s3.Bucket('mybucket')
+        with self.stubber:
+            response = bucket.copy(self.copy_source, 'mykey')
+        # The response will be none on a successful transfer.
+        self.assertIsNone(response)
+
+    def test_object_copy(self):
+        obj = self.s3.Object('mybucket', 'mykey')
+        with self.stubber:
+            response = obj.copy(self.copy_source)
+        self.assertIsNone(response)
```

Comment:
It also may be good to check that the progress subscribers work as well in a functional test. I would avoid making it a unit test because you would have to patch S3Transfer.